File: task4laba6.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <iomanip>

using namespace std;

int N, M;

int encode(vector<vector<int> >& board)
{
    int mask = 0;
    int pos = 0;

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            if (board[i][j] == 1)
            {
                mask |= (1 << pos);
            }

            pos++;
        }
    }
    return mask;
}

vector<vector<int>> decode(int mask)
{
    vector<vector<int>> board(N, vector<int>(M, 0));
    int pos = 0;

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            board[i][j] = (mask >> pos) & 1;
            pos++;
        }
    }
    return board;
}

int buildTarget(int edgeVal)
{
    vector<vector<int>> target(N, vector<int>(M, 0));

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            if (i == 0 || i == N-1 || j == 0 || j == M-1)
            {
                target[i][j] = edgeVal;
            }
            else
            {
                target[i][j] = 1 - edgeVal;
            }
        }
    }
    return encode(target);
}

int bfs(int startMask, int targetMask)
{
    if (startMask == targetMask)
    {
        return 0;
    }

    unordered_map<int,int> visited;
    queue<int> q;

    visited[startMask] = 0;
    q.push(startMask);

    int dx[4] = {0, 0, 1, -1};
    int dy[4] = {1, -1, 0, 0};

    while (!q.empty())
    {
        int cur = q.front();
        q.pop();

        int curDist = visited[cur];

        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < M; j++)
            {
                for (int d = 0; d < 4; d++)
                {
                    int ni = i + dx[d];
                    int nj = j + dy[d];

                    if (ni < 0 || ni >= N || nj < 0 || nj >= M)
                    {
                        continue;
                    }

                    int pos1 = i  * M + j;
                    int pos2 = ni * M + nj;

                    int newMask = cur ^ (1 << pos1) ^ (1 << pos2);

                    if (newMask == targetMask)
                    {
                        return curDist + 1;
                    }

                    if (visited.count(newMask) == 0)
                    {
                        visited[newMask] = curDist + 1;
                        q.push(newMask);
                    }
                }
            }
        }
    }

    return -1;
}

int main()
{
    cout << "Введите N (от 2 до 4): ";

    while (!(cin >> N) || N < 2 || N > 4)
    {
        cout << "Ошибка: введите от 2 до 4\n";
        cin.clear();
        cin.ignore(1000, '\n');
        cout << "Введите N (от 2 до 4): ";
    }

    cout << "Введите M (от 2 до 4): ";

    while (!(cin >> M) || M < 2 || M > 4)
    {
        cout << "Ошибка: введите от 2 до 4\n";
        cin.clear();
        cin.ignore(1000, '\n');
        cout << "Введите M (от 2 до 4): ";
    }

    vector<vector<int>> board(N, vector<int>(M, 0));

    cout << "Введите доску (0 и 1):\n";

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            while (!(cin >> board[i][j]) || board[i][j] < 0 || board[i][j] > 1)
            {
                cout << "Ошибка: введите 0 или 1\n";
                cin.clear();
                cin.ignore(1000, '\n');
            }
        }
    }

    cout << "\nВведённая доска:\n";

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            cout << setw(3) << board[i][j];
        }
        cout << "\n";
    }

    int startMask = encode(board);
    int bestAnswer = 1000000;

    for (int edgeVal = 0; edgeVal < 2; edgeVal++)
    {
        int targetMask = buildTarget(edgeVal);

        vector<vector<int>> targetBoard = decode(targetMask);

        cout << "\nЦелевая рамка (края=" << edgeVal << ", внутри=" << (1 - edgeVal) << "):\n";

        for (int i = 0; i < N; i++)
        {
            for (int j = 0; j < M; j++)
            {
                cout << setw(3) << targetBoard[i][j];
            }
            cout << "\n";
        }

        int result = bfs(startMask, targetMask);

        if (result == -1)
        {
            cout << "Невозможно достичь.\n";
        }
        else
        {
            cout << "Ходов: " << result << "\n";

            if (result < bestAnswer)
            {
                bestAnswer = result;
            }
        }
    }

    cout << "\n" << setw(30) << left << "Минимальное число ходов:" << bestAnswer << "\n";

    return 0;
}

File: Task4laba6.rs
use std::collections::{HashMap, VecDeque};
use std::io;

static mut N: usize = 0;
static mut M: usize = 0;

fn encode(board: &Vec<Vec<i32>>) -> i32 {
    let mut mask = 0;
    let mut pos = 0;

    unsafe {
        for i in 0..N {
            for j in 0..M {
                if board[i][j] == 1 {
                    mask |= 1 << pos;
                }
                pos += 1;
            }
        }
    }

    mask
}

fn decode(mask: i32) -> Vec<Vec<i32>> {
    let mut board;

    unsafe {
        board = vec![vec![0; M]; N];

        let mut pos = 0;

        for i in 0..N {
            for j in 0..M {
                board[i][j] = (mask >> pos) & 1;
                pos += 1;
            }
        }
    }

    board
}

fn build_target(edge_val: i32) -> i32 {
    let mut target;

    unsafe {
        target = vec![vec![0; M]; N];

        for i in 0..N {
            for j in 0..M {
                if i == 0 || i == N - 1 || j == 0 || j == M - 1 {
                    target[i][j] = edge_val;
                } else {
                    target[i][j] = 1 - edge_val;
                }
            }
        }
    }

    encode(&target)
}

fn bfs(start_mask: i32, target_mask: i32) -> i32 {
    if start_mask == target_mask {
        return 0;
    }

    let mut visited: HashMap<i32, i32> = HashMap::new();
    let mut q: VecDeque<i32> = VecDeque::new();

    visited.insert(start_mask, 0);
    q.push_back(start_mask);

    let dx = [0, 0, 1, -1];
    let dy = [1, -1, 0, 0];

    while !q.is_empty() {
        let cur = q.pop_front().unwrap();
        let cur_dist = *visited.get(&cur).unwrap();

        unsafe {
            for i in 0..N {
                for j in 0..M {
                    for d in 0..4 {
                        let ni = i as i32 + dx[d];
                        let nj = j as i32 + dy[d];

                        if ni < 0 || ni >= N as i32 || nj < 0 || nj >= M as i32 {
                            continue;
                        }

                        let pos1 = (i * M + j) as i32;
                        let pos2 = (ni as usize * M + nj as usize) as i32;

                        let new_mask = cur ^ (1 << pos1) ^ (1 << pos2);

                        if new_mask == target_mask {
                            return cur_dist + 1;
                        }

                        if !visited.contains_key(&new_mask) {
                            visited.insert(new_mask, cur_dist + 1);
                            q.push_back(new_mask);
                        }
                    }
                }
            }
        }
    }

    -1
}

fn read_number() -> i32 {
    loop {
        let mut s = String::new();
        io::stdin().read_line(&mut s).unwrap();

        if let Ok(x) = s.trim().parse::<i32>() {
            return x;
        }

        println!("Ошибка ввода.");
    }
}

fn main() {
    unsafe {
        println!("Введите N (от 2 до 4): ");

        loop {
            let x = read_number();

            if x >= 2 && x <= 4 {
                N = x as usize;
                break;
            }

            println!("Ошибка: введите от 2 до 4");
        }

        println!("Введите M (от 2 до 4): ");

        loop {
            let x = read_number();

            if x >= 2 && x <= 4 {
                M = x as usize;
                break;
            }

            println!("Ошибка: введите от 2 до 4");
        }

        let mut board = vec![vec![0; M]; N];

        println!("Введите доску (0 и 1):");

        for i in 0..N {
            for j in 0..M {
                loop {
                    let x = read_number();

                    if x == 0 || x == 1 {
                        board[i][j] = x;
                        break;
                    }

                    println!("Ошибка: введите 0 или 1");
                }
            }
        }

        println!("\nВведённая доска:");

        for i in 0..N {
            for j in 0..M {
                print!("{:3}", board[i][j]);
            }
            println!();
        }

        let start_mask = encode(&board);
        let mut best_answer = 1000000;

        for edge_val in 0..2 {
            let target_mask = build_target(edge_val);
            let target_board = decode(target_mask);

            println!(
                "\nЦелевая рамка (края={}, внутри={}):",
                edge_val,
                1 - edge_val
            );

            for i in 0..N {
                for j in 0..M {
                    print!("{:3}", target_board[i][j]);
                }
                println!();
            }

            let result = bfs(start_mask, target_mask);

            if result == -1 {
                println!("Невозможно достичь.");
            } else {
                println!("Ходов: {}", result);

                if result < best_answer {
                    best_answer = result;
                }
            }
        }

        println!("\nМинимальное число ходов: {}", best_answer);
    }
}